README.md
# satinv - Satellite Inventory

This program will generate an [Ansible](https://ansible.com) [Dynamic Inventory](https://docs.ansible.com/ansible/latest/user_guide/intro_dynamic_inventory.html) using [Red Hat Satellite](https://www.redhat.com/en/technologies/management/satellite) as a source.  The Dynamic Inventory will consist of a number of inventory groups:-
* Host Collections - Within Red Hat Satellite, hosts can be assigned to one or more Host Collections.
* Valid Hosts - In this context a host is valid if it has:
    * an IP address (ipv4 or ipv6)
    * a valid Red Hat subscription
    * been seen within an acceptable period of time - 7 days default (config option: sat_valid_days)
* CIDR groups - Groups created based on the subnet the host resides in

## Installation
* Grab a copy of Google's [Go](https://golang.org/) and follow the instructions for your platform to install it.
* Download the **satinv** repository and compile it.
* Copy the resulting binary to somewhere sane (on Linux, /usr/local/bin is probably a good choice).
* Try executing `satinv --help` to check you don't have any runtime errors.

## Configuration
The configuration for **satinv** lives in a single YAML formatted file.  The file can be located anywhere but the default is `/etc/ansible/satinv.yml`.
The location can be overridden with `--config=/path/to/config.yml` or by setting the environment variable `SATINVCFG`.  **Note**: You cannot use the --config option when running satinv from `ansible-playbook` or `ansible-inventory`.  This is a constraint imposed by Ansible.  A basic config file could be:
```
---
api:
  baseurl: https://www.redhat.mydomain.com
  user: myreadonlyuser
  password: myPassword

cache:
  dir: ~/satinv/cache
  validity: 28800

target_filename: ~/satinv/inventory.json

cidrs:
  dev: 192.168.0.0/24
  test: 192.168.1.0/24
  prod: 192.168.100.0/23
```
Test the configuration by running `satinv --debug` (assuming your config path is predefined).

## Usage
To use the dynamic inventory consider the following commands:
* `ansible-inventory -i /usr/local/bin/satinv --list`
* `ansible-playbook -i /usr/local/bin/satinv my_playbook.yml`

CHANGELOG
2021-11-28 v0.0.4
    * Add caching of inventory output
    * Only write the expiry file once per run

2021-11-17 v0.0.3
    * Add a CHANGELOG.
    * Create config sections for API and Cache settings
    * Add support for homedir (~) expansion in paths
    * Remove redundant use_homedir config option (see above change)

2021-10-21 v0.0.2
    * Add a use_homedir option to make paths relative to homedir

2021-09-01 v0.0.1
    * Initial release


cidrs/cidrs_test.go
package cidrs

import (
	"testing"
)

func contains(sl []string, st string) bool {
	for _, sls := range sl {
		if sls == st {
			return true
		}
	}
	return false
}

func TestCIDR(t *testing.T) {
	c := make(Cidrs)
	c.AddCIDRMap(map[string]string{
		"test1": "192.168.0.0/24",
		"test2": "192.168.1.0/24",
	})
	testAddr := "192.168.0.5"
	memberOf := c.ParseCIDRs(testAddr)
	if !contains(memberOf, "test1") {
		t.Fatalf("%s should be a member of test1", testAddr)
	}
	if contains(memberOf, "test2") {
		t.Fatalf("%s should not be a member of test2", testAddr)
	}

	testAddr = "192.168.1.254"
	memberOf = c.ParseCIDRs(testAddr)
	if !contains(memberOf, "test2") {
		t.Fatalf("%s should be a member of test2", testAddr)
	}
	if contains(memberOf, "test1") {
		t.Fatalf("%s should not be a member of test1", testAddr)
	}
}


cacher/cacher_test.go
package cacher

import (
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path"
	"strings"
	"testing"
	"time"

	"github.com/tidwall/gjson"
)

func mkTempDir() string {
	tempDir, err := ioutil.TempDir("/tmp", "sat")
	if err != nil {
		log.Fatalf("Unable to create TempDir: %v", err)
	}
	return tempDir
}

// abs implements the standard abs function
func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

func TestCacher(t *testing.T) {
	tempDir := mkTempDir()
	defer os.RemoveAll(tempDir)
	c := NewCacher(tempDir)
	if c.cacheDir != tempDir {
		t.Fatalf("Unexpected cacheDir.  Expected=%s, Got=%s", tempDir, c.cacheDir)
	}
}

func TestExpire(t *testing.T) {
	c := NewCacher("fake")
	var durationSecs int64 = 2
	c.SetCacheDuration(durationSecs)
	now := time.Now().Unix()
	expire := expireTime(c.cachePeriod)
	if abs(expire-(now+durationSecs)) > 2 {
		t.Fatalf("Unexpected cachePeriod.  Expected=%d, Got=%d", expire, now)
	}
	if time.Now().Unix() > expire {
		t.Fatal("Cache should not be expired")
	}
	time.Sleep(time.Second * time.Duration(3))
	if time.Now().Unix() < expire {
		t.Fatal("Cache should be expired")
	}
}

func TestWrite(t *testing.T) {
	tempDir := mkTempDir()
	defer os.RemoveAll(tempDir)
	c := NewCacher(tempDir)
	sample := `{"results": ["a","b","c"]}`
	j := gjson.Parse(sample)
	c.jsonToFile(path.Join(tempDir, "test.json"), j)
}

func TestGetURL(t *testing.T) {
	tempDir := mkTempDir()
	defer os.RemoveAll(tempDir)
	c := NewCacher(tempDir)
	c.SetCacheDuration(60)
	testURL := "http://fakeurl.fake"
	testFile := "test.json"
	_, err := c.GetURL(testURL)
	if err == nil {
		t.Fatalf("No error returned for non existent cache file")
	}
	c.AddURL(testURL, testFile)
	_, err = c.GetURL(testURL)
	if !errors.Is(err, errAPIInit) {
		t.Fatalf("Error: %v", err)
	}
}

func TestExportExpiry(t *testing.T) {
	tempDir := mkTempDir()
	defer os.RemoveAll(tempDir)
	c := NewCacher(tempDir)
	testItem := "http://fakeurl.fake"
	testFile := "test.json"
	c.AddURL(testItem, testFile)
	// UpdateExpiry also writes the JSON to the test.json file.
	testExpire := c.UpdateExpiry(testItem, c.cachePeriod)
	c.WriteExpiryFile()

	// Create an empty file for the cache item.  This prevents HasExpired from returning true due to the absense of
	// the file.
	fullTestFile := path.Join(tempDir, testFile)
	emptyFile, err := os.Create(fullTestFile)
	if err != nil {
		log.Fatal(err)
	}
	emptyFile.Close()

	// Test the cache item filename matches testFile
	if c.GetFilename(testItem) != fullTestFile {
		t.Errorf("Unexpected filename: Expected=%s, Got=%s", fullTestFile, c.GetFilename(testItem))
	}

	// Fetch the JSON from tempfile
	filename := path.Join(tempDir, cacheExpiryFile)
	gj, err := c.jsonFromFile(filename)
	if err != nil {
		t.Fatal(err)
	}
	// For gjson, the dot needs to be escaped.
	gjSafeItem := strings.Replace(testItem, ".", "\\.", -1)
	gjItem := fmt.Sprintf("urls.%s", gjSafeItem)
	//getURL := "urls.http://fakeurl\\.fake"
	getItem := gj.Get(gjItem)
	if !getItem.Exists() {
		t.Errorf("Expected url key does not exist: %s", testItem)
	}
	getExpire := getItem.Int()
	if getExpire != testExpire {
		t.Fatalf("Unexpected expiry value.  Expected=%d, Got=%d", testExpire, getExpire)
	}

	// Test HasExpired.
	expiry, err := c.HasExpired(testItem)
	if err != nil {
		log.Fatal(err)
	}
	if expiry {
		t.Error("Item cache should not be expired")
	}
	// Reset the item cache expiry to epoch zero (expired)
	c.cacheExpiry[testItem] = 0
	expiry, err = c.HasExpired(testItem)
	if err != nil {
		log.Fatal(err)
	}
	if !expiry {
		t.Error("Item cache should be expired")
	}
}


config/config_test.go
package config

import (
	"io/ioutil"
	"os"
	"os/user"
	"path"
	"testing"
)

func TestFlags(t *testing.T) {
	expectedConfig := "/etc/promsat/fake.yml"
	// This needs to be set prior to doing ParseFlags()
	os.Setenv("SATINVCFG", expectedConfig)
	f := ParseFlags()
	if f.Config != expectedConfig {
		t.Fatalf("Expected --config to contain \"%v\" but got \"%v\".", expectedConfig, f.Config)
	}
}

func TestConfig(t *testing.T) {
	testFile, err := ioutil.TempFile("/tmp", "testcfg")
	if err != nil {
		t.Fatalf("Unable to create TempFile: %v", err)
	}
	defer os.Remove(testFile.Name())
	fakeCfg := new(Config)
	fakeCfg.SatValidDays = defaultSatValidDays
	fakeCfg.Cache.Validity = defaultCacheValiditySeconds
	fakeCfg.Cache.InventoryValidity = defaultInventoryValiditySeconds
	fakeCfg.InventoryPrefix = "sat_"
	fakeCfg.WriteConfig(testFile.Name())

	cfg, err := ParseConfig(testFile.Name())
	if err != nil {
		t.Fatalf("ParseConfig returned: %v", err)
	}

	if cfg.SatValidDays != fakeCfg.SatValidDays || cfg.SatValidDays != defaultSatValidDays {
		t.Fatalf(
			"Unexpected config.SatValidDays. Default=%d, Expected=%d, Got=%d",
			defaultSatValidDays, fakeCfg.SatValidDays, cfg.SatValidDays)
	}
	if cfg.Cache.Validity != fakeCfg.Cache.Validity || cfg.Cache.Validity != defaultCacheValiditySeconds {
		t.Fatalf(
			"Unexpected config.Cache.Validity. Default=%d, Expected=%d, Got=%d",
			defaultCacheValiditySeconds, fakeCfg.Cache.Validity, cfg.Cache.Validity)
	}
	if cfg.Cache.InventoryValidity != fakeCfg.Cache.InventoryValidity || cfg.Cache.InventoryValidity != defaultInventoryValiditySeconds {
		t.Fatalf(
			"Unexpected config.Cache.InventoryValidity. Default=%d, Expected=%d, Got=%d",
			defaultInventoryValiditySeconds, fakeCfg.Cache.InventoryValidity, cfg.Cache.InventoryValidity)
	}
	if cfg.InventoryPrefix != fakeCfg.InventoryPrefix {
		t.Errorf(
			"Unexpected InventoryPrefix. Expected=%s, Got=%s", fakeCfg.InventoryPrefix, cfg.InventoryPrefix)
	}
}

func TestExpandTilde(t *testing.T) {
	u, err := user.Current()
	if err != nil {
		t.Fatalf("Unable to ascertain current user: %v", err)
	}
	// Test homedir, without path
	testDir := "~"
	expectDir := u.HomeDir
	resultDir := expandTilde(testDir)
	if expectDir != resultDir {
		t.Errorf("Tilde expansion failed.  Expected=%s, Got=%s", expectDir, resultDir)
	}
	// Test homedir with path
	testDir = "~/dir1/dir2/filename"
	expectDir = path.Join(u.HomeDir, "dir1/dir2/filename")
	resultDir = expandTilde(testDir)
	if expectDir != resultDir {
		t.Errorf("Tilde expansion failed.  Expected=%s, Got=%s", expectDir, resultDir)
	}
	// Test path without homedir
	testDir = "/dir1/dir2/filename"
	expectDir = "/dir1/dir2/filename"
	resultDir = expandTilde(testDir)
	if expectDir != resultDir {
		t.Errorf("Tilde expansion failed.  Expected=%s, Got=%s", expectDir, resultDir)
	}
}


satinv.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strings"
	"time"

	"github.com/crooks/satinv/cacher"
	"github.com/crooks/satinv/cidrs"
	"github.com/crooks/satinv/config"
	"github.com/tidwall/gjson"
	"github.com/tidwall/sjson"
)

const (
	inventoryName string = "inventory"
)

var (
	cfg   *config.Config
	flags *config.Flags
)

type inventory struct {
	json  string
	cache *cacher.Cache
}

// shortName take a hostname string and returns the shortname for it.
func shortName(host string) string {
	return strings.Split(host, ".")[0]
}

// satTimestamp parses a DateTime string of the format used in the Satellite API
func satTimestamp(ts string) (t time.Time, err error) {
	layout := "2006-01-02 15:04:05 MST"
	t, err = time.Parse(layout, ts)
	if err != nil {
		log.Printf("Sat time parse: %v", err)
		return
	}
	return
}

// getHostByID returns the string representation of a hostname for a given ID string.
func getHostByID(hosts gjson.Result, id string) (string, error) {
	query := fmt.Sprintf("results.#(id=\"%s\").name", id)
	hostname := hosts.Get(query)
	if hostname.Exists() {
		return hostname.String(), nil
	}
	err := fmt.Errorf("name not found for id: %s", id)
	return "", err
}

// mkInventoryName converts a Host Collection name to something compatible with Ansible Inventories.
func mkInventoryName(s string) string {
	s = strings.ToLower(s)
	s = strings.ReplaceAll(s, " ", "_")
	s = cfg.InventoryPrefix + s
	return s
}

// getHostCollection takes an ID string and returns the Host Collection associated with it.
func (inv *inventory) getHostCollection(id string) (gjson.Result, error) {
	collectionURL := fmt.Sprintf("%s/katello/api/host_collections/%s", cfg.API.BaseURL, id)
	collectionFilename := fmt.Sprintf("host_collections_%s.json", id)
	inv.cache.AddURL(collectionURL, collectionFilename)
	collection, err := inv.cache.GetURL(collectionURL)
	if err != nil {
		return gjson.Result{}, err
	}
	collectionID := collection.Get("id")
	if !collectionID.Exists() {
		err := errors.New("host collection has no ID field")
		return gjson.Result{}, err
	}
	if collectionID.String() != id {
		err := errors.New("host collection ID does not match requested ID")
		return gjson.Result{}, err
	}
	return collection, nil
}

// importCIDRs constructs a new instance of Cidrs and then populates it from a map in the Config.
func importCIDRs() cidrs.Cidrs {
	cidr := make(cidrs.Cidrs)
	cidr.AddCIDRMap(cfg.CIDRs)
	return cidr
}

// refreshInventory produces a new inventory.json copy from the Satellite API (or cache).
func (inv *inventory) refreshInventory() {
	// If URLs have to be pulled from an API, this has to be initialised.
	inv.cache.InitAPI(cfg.API.User, cfg.API.Password, cfg.API.CertFile)

	// Populate the hosts object
	hostsURL := fmt.Sprintf("%s/api/v2/hosts?per_page=1000", cfg.API.BaseURL)
	inv.cache.AddURL(hostsURL, "hosts.json")
	hosts, err := inv.cache.GetURL(hostsURL)
	if err != nil {
		log.Fatalf("Unable to read hosts from JSON file: %v", err)
	}

	// Initialize the inventory object that contains the json string field
	inv.json = "{}"
	inv.json, err = sjson.Set(inv.json, "_meta", "hostvars")
	if err != nil {
		log.Fatal(err)
	}
	inv.parseHosts(hosts)
	inv.parseHostCollections(hosts)
	// For human readability, put an LF on the end of the json.
	inv.json += "\n"
	filename := inv.cache.GetFilename(inventoryName)
	err = ioutil.WriteFile(filename, []byte(inv.json), 0644)
	if err != nil {
		log.Fatalf("WriteFile: %v", err)
	}
	// If the inventory has been successfully refreshed, update the expiry file with a new refresh timestamp.
	// TODO: Make the inventory cache duration configurable
	inv.cache.UpdateExpiry(inventoryName, 3600)
}

// mkInventory assembles all the components of a Dynamic Inventory and writes them to Stdout (or a file).
func mkInventory() {
	// Initialize an inventory struct
	inv := new(inventory)
	// Initialize the URL cache
	inv.cache = cacher.NewCacher(cfg.Cache.Dir)
	// When this function completes, write the expiry file (if one or more cache items have been refreshed).
	defer inv.cache.WriteExpiryFile()
	inv.cache.SetCacheDuration(cfg.Cache.Validity)
	if flags.Refresh {
		// Force a cache refresh
		inv.cache.SetRefresh()
	}

	// This isn't a real URL; it never gets pulled from an API.  It contains the output inventory JSON and enables it
	// to be cached.
	inv.cache.AddURL(inventoryName, fmt.Sprintf("%s.json", inventoryName))
	refresh, err := inv.cache.HasExpired(inventoryName)
	if err != nil {
		log.Fatal(err)
	}
	if refresh {
		inv.refreshInventory()
	} else {
		inv.json = string(inv.cache.GetFile(inventoryName))
	}
	if flags.List {
		_, err = fmt.Fprint(os.Stdout, inv.json)
		if err != nil {
			log.Fatalf("Fprintf: %v", err)
		}
	}
}

// parseHosts creates the inventory hostvars metadata for each host
func (inv *inventory) parseHosts(hosts gjson.Result) {
	defer timeTrack(time.Now(), "parseHosts")
	var err error

	// Import the CIDRs we want to test each address against.
	cidr := importCIDRs()
	if len(cidr) == 0 {
		log.Print("Bypassing CIDR membership processing.  No CIDRs defined.")
	}

	// Initialize the sat_valid inventory group
	// satValidAppend is a special string used by sjson to append entries to an inventory group
	satValidAppend := fmt.Sprintf("%svalid.hosts.-1", cfg.InventoryPrefix)
	// Add sat_valid to the all{children} array
	inv.json, err = sjson.Set(inv.json, "all.children.-1", cfg.InventoryPrefix+"valid")
	if err != nil {
		log.Fatal(err)
	}

	// Iterate through each host in the Satellite results
	for _, h := range hosts.Get("results").Array() {
		// Every individual host map should contain a "name" key
		if !h.Get("name").Exists() {
			log.Print("No hostname found in Satellite host map")
			continue
		}
		hostNameShort := shortName(h.Get("name").String())
		key := fmt.Sprintf("_meta.hostvars.%s", hostNameShort)
		inv.json, err = sjson.Set(inv.json, key, h.Value())
		if err != nil {
			log.Fatal(err)
		}
		inv.hgSatValid(h, satValidAppend, hostNameShort)
		if len(cidr) > 0 {
			inv.hgCIDRMembers(h, cidr)
		}
	}
}

// parseHostCollections iterates through the Satellite Host Collections and associates hostnames with the each
// Collection's host_ids.
func (inv *inventory) parseHostCollections(hosts gjson.Result) {
	defer timeTrack(time.Now(), "parseHostCollections")
	collectionsURL := fmt.Sprintf("%s/katello/api/host_collections", cfg.API.BaseURL)
	inv.cache.AddURL(collectionsURL, "host_collections.json")
	collections, err := inv.cache.GetURL(collectionsURL)
	if err != nil {
		log.Fatalf("Unable to read JSON from file: %v", err)
	}
	for _, c := range collections.Get("results").Array() {
		hostCollectionName := c.Get("name").String()
		hostCollectionID := c.Get("id").String()
		hostCollection, err := inv.getHostCollection(hostCollectionID)
		if err != nil {
			log.Printf("Unable to get host_collection: %v", err)
			continue
		}
		collectionKey := mkInventoryName(hostCollectionName)
		inv.json, err = sjson.Set(inv.json, "all.children.-1", collectionKey)
		if err != nil {
			log.Fatal(err)
		}
		collectionAppend := fmt.Sprintf("%s.hosts.-1", collectionKey)
		for _, v := range hostCollection.Get("host_ids").Array() {
			host, err := getHostByID(hosts, v.String())
			if err != nil {
				log.Printf("Cannot fetch host by ID: %v", err)
				continue
			}
			inv.json, err = sjson.Set(inv.json, collectionAppend, shortName(host))
			if err != nil {
				log.Fatal(err)
			}
		}
	}
}

// satValid creates an inventory group of hosts the meet "valid" conditions.
func (inv *inventory) hgSatValid(host gjson.Result, satValidAppend, hostNameShort string) {
	// Check the host has an IP address
	ip4 := host.Get("ip")
	ip6 := host.Get("ip6")
	if !ip4.Exists() && !ip6.Exists() {
		log.Printf("satValid: No IP address found for %s", hostNameShort)
		return
	}
	if ip4.String() == "" && ip6.String() == "" {
		log.Printf("satValid: No valid IP addresses found for %s", hostNameShort)
		return
	}
	// Ensure the host has a valid subscription
	subStatus := host.Get("subscription_status")
	if !subStatus.Exists() || subStatus.Int() != 0 {
		log.Printf("satValid: subscription_status not found for %s", hostNameShort)
		return
	}
	if subStatus.Int() != 0 {
		log.Printf("satValid: Invalid subscription status (%d) for %s", subStatus.Int(), hostNameShort)
		return
	}

	// Check last_checkin date
	checkin := host.Get("subscription_facet_attributes.last_checkin")
	if !checkin.Exists() {
		log.Printf("satValid: subscription_facet_attributes.last_checkin not found for %s", hostNameShort)
		return
	}
	satTime, err := satTimestamp(checkin.String())
	if err != nil {
		// consider the host to be invalid
		log.Printf("satValid: Invalid date/time %s for %s", checkin.String(), hostNameShort)
		return
	}
	oldestValidTime := time.Now().Add(-time.Hour * 24 * time.Duration(cfg.SatValidDays))
	if satTime.Before(oldestValidTime) {
		log.Printf("satValid: Last checkin for %s is too old", hostNameShort)
		return
	}

	// All the above conditions passed; this is a valid host.
	inv.json, err = sjson.Set(inv.json, satValidAppend, hostNameShort)
	if err != nil {
		log.Fatal(err)
	}
}

// hgCIDRMembers compares the IPv4 address of the current host to a list of CIDRs.  When the address is a member of a
// CIDR, its appended to an inventory group for that CIDR.
func (inv *inventory) hgCIDRMembers(host gjson.Result, cidr cidrs.Cidrs) {
	hostNameShort := shortName(host.Get("name").String())

	// Test the validity of the address for CIDR membership processing.
	gjIP4 := host.Get("ip")
	if !gjIP4.Exists() {
		return
	}
	ip4 := gjIP4.String()
	if ip4 == "" {
		return
	}

	// invGrps will contain a slice of all inventory groups the address is a member of.
	invGrps := cidr.ParseCIDRs(ip4)

	var err error
	for _, invGrp := range invGrps {
		sjKey := fmt.Sprintf("%s.hosts.-1", mkInventoryName(invGrp))
		inv.json, err = sjson.Set(inv.json, sjKey, hostNameShort)
		if err != nil {
			log.Printf("hgCIDRMembers: %s: %v", hostNameShort, err)
		}
	}
}

// timeTrack can be used to time the processing duration of a function.
func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

func main() {
	var err error
	flags = config.ParseFlags()
	// For normal use, log output needs to be surpressed or inventory's dynamic inventory will try to process it.
	if !flags.Debug {
		log.SetOutput(ioutil.Discard)
	}
	cfg, err = config.ParseConfig(flags.Config)
	if err != nil {
		log.Fatalf("Cannot parse config: %v", err)
	}

	mkInventory()
}


cidrs/cidr.go
// cidrs provides a method for a testing membership of a given IPv4 address to multiple subnets.
package cidrs

import (
	"fmt"
	"log"
	"net"
)

// Cidrs contains a map of desired inventory groups and the subnets associated with them
type Cidrs map[string]*net.IPNet

// parseCIDRs compares an IP address to a range of subnets.  If the address is in the subnet, the name of the subnet
// is appended to a subnets list and returned
func (c Cidrs) ParseCIDRs(ipAddr string) (memberOf []string) {
	cidrString := fmt.Sprintf("%s/32", ipAddr)
	ip, _, err := net.ParseCIDR(cidrString)
	if err != nil {
		log.Print("Invalid CIDR address")
	}

	// Iterate through each defined subnet and test if the address is a member of it.
	for name, cidr := range c {
		if cidr.Contains(ip) {
			memberOf = append(memberOf, name)
		}
	}
	return
}

// AddCIDR adds a CIDR name and subnet to the members list.
func (c Cidrs) AddCIDR(name, subnet string) {
	_, cidr, err := net.ParseCIDR(subnet)
	if err != nil {
		log.Printf("Invalid subnet: %v", err)
	}
	c[name] = cidr
}

// AddCIDRMap is a helper function for adding multiple subnets to the members list.
func (c Cidrs) AddCIDRMap(cidrMap map[string]string) {
	for name, cidr := range cidrMap {
		c.AddCIDR(name, cidr)
	}
}


cacher/satapi/satapi.go
// satapi provides an API interface to Red Hat Satellite
package satapi

import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
)

// AuthClient contains the HTTP client components
type AuthClient struct {
	Username   string
	Password   string
	HTTPClient *http.Client
}

// NewBasicAuthClient returns an instance of AuthClient
func NewBasicAuthClient(username, password, certFile string) *AuthClient {
	return &AuthClient{
		Username:   username,
		Password:   password,
		HTTPClient: httpAuthClient(certFile),
	}
}

// GetJSON takes a URL relating to a Rest API and returns the resulting JSON as a byte slice.
func (s *AuthClient) GetJSON(url string) ([]byte, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	bytes, err := s.doRequest(req)
	if err != nil {
		return nil, err
	}
	return bytes, nil
}

// httpAuthClient creates a new instance of http.Client with support for
// additional rootCAs.  As XClarity is frequently installed as an appliance,
// with a self-signed cert, this appears to be quite useful.
func httpAuthClient(certFile string) *http.Client {
	rootCAs, err := x509.SystemCertPool()
	if err != nil {
		log.Fatal(err)
	}
	if rootCAs == nil {
		rootCAs = x509.NewCertPool()
	}
	certs, err := ioutil.ReadFile(certFile)
	if errors.Is(err, os.ErrNotExist) {
		//log.Println("No additional certificates imported")
	} else if err != nil {
		panic(err)
	} else if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
		log.Println("Cert import failed.  Proceeding with system CAs.")
	}
	config := &tls.Config{
		InsecureSkipVerify: false,
		RootCAs:            rootCAs,
	}
	tr := &http.Transport{TLSClientConfig: config}
	return &http.Client{Transport: tr}
}

// doRequest does an HTTP URL request and returns it as a byte array
func (s *AuthClient) doRequest(req *http.Request) ([]byte, error) {
	req.SetBasicAuth(s.Username, s.Password)
	resp, err := s.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("Status error: %s\n", string(body))
	}
	return body, nil
}


config/config.go
// config provides flag and configuration file reading for satinv
package config

import (
	"flag"
	"io/ioutil"
	"os"
	"os/user"
	"path"
	"strings"

	"gopkg.in/yaml.v2"
)

const (
	defaultSatValidDays             int   = 7
	defaultCacheValiditySeconds     int64 = 8 * 60 * 60 // 8 Hours
	defaultInventoryValiditySeconds int64 = 2 * 60 * 60 // 2 Hours
)

// Config contains all the configuration settings
type Config struct {
	API struct {
		BaseURL  string `yaml:"baseurl"`
		CertFile string `yaml:"certfile"`
		Password string `yaml:"password"`
		User     string `yaml:"user"`
	} `yaml:"api"`
	Cache struct {
		Dir               string `yaml:"dir"`
		Validity          int64  `yaml:"validity"`
		InventoryValidity int64  `yaml:"inventory_validity"`
	} `yaml:"cache"`
	CIDRs           map[string]string `yaml:"cidrs"`
	InventoryPrefix string            `yaml:"inventory_prefix"`
	SatValidDays    int               `yaml:"sat_valid_days"`
}

// Flags are the command line flags
type Flags struct {
	Config  string
	Debug   bool
	List    bool
	Refresh bool
}

// WriteConfig will create a YAML formatted config file from a Config struct
func (c *Config) WriteConfig(filename string) error {
	data, err := yaml.Marshal(c)
	if err != nil {
		return err
	}
	err = ioutil.WriteFile(filename, data, 0644)
	if err != nil {
		return err
	}
	return nil
}

// ParseFlags transcribes command line flags into a struct
func ParseFlags() *Flags {
	f := new(Flags)
	// Config file
	flag.StringVar(&f.Config, "config", "", "Config file")
	flag.BoolVar(&f.Debug, "debug", false, "Write logoutput to stderr")
	flag.BoolVar(&f.List, "list", false, "Produce a full inventory to stdout")
	flag.BoolVar(&f.Refresh, "refresh", false, "Force a cache refresh")
	flag.Parse()

	// If a "--config" flag has been provided, it should be honoured (even if it's invalid or doesn't exist).
	if f.Config == "" {
		if os.Getenv("SATINVCFG") == "" {
			// Environment variable hasn't been set.  No options left so take a bold guess at a config location.
			f.Config = "/etc/ansible/satinv.yml"
		} else {
			// Assume the SATINVCFG variable contains something meaningful and valid.
			f.Config = os.Getenv("SATINVCFG")
		}
	}
	return f
}

// ParseConfig expects a YAML formatted config file and populates a Config struct
func ParseConfig(filename string) (*Config, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	y := yaml.NewDecoder(file)
	config := new(Config)
	// Read the config file
	if err := y.Decode(&config); err != nil {
		return nil, err
	}
	// Set config defaults here
	if config.SatValidDays == 0 {
		config.SatValidDays = defaultSatValidDays
	}
	if config.Cache.Validity == 0 {
		config.Cache.Validity = defaultCacheValiditySeconds
	}
	if config.Cache.InventoryValidity == 0 {
		config.Cache.InventoryValidity = defaultInventoryValiditySeconds
	}

	// The following config options may need tilde expansion
	config.Cache.Dir = expandTilde(config.Cache.Dir)

	return config, nil
}

// expandTilde expands filenames and paths that use the tilde convention to imply relative to homedir.
func expandTilde(inPath string) (outPath string) {
	u, err := user.Current()
	if err != nil {
		panic(err)
	}
	if inPath == "~" {
		outPath = u.HomeDir
	} else if strings.HasPrefix(inPath, "~/") {
		outPath = path.Join(u.HomeDir, inPath[2:])
	} else {
		outPath = inPath
	}
	return
}
